Bug: [500 Internal Server Error] Input/output error (os error 5) with Azure Blob Fuse

### What happened?

I'm using [ReductStore Azure VM](https://azuremarketplace.microsoft.com/en-us/marketplace/apps/reductsoftware.reductstore-server?tab=Overview) with an Azure storage account.  I've encountered input/output errors when trying to copy data from one bucket to another from the local host:

```
reduct-cli cp local/datasets local/demo --limit 1000
[00:00:09, ETA 00:00:00] █████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 23.300% [InternalServerError] Input/output error (os error 5)
[00:00:11, ETA 00:00:00] ███████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 29.100% [InternalServerError] Input/output error (os error 5)
[00:00:13, ETA 00:00:00] ████████████████████████████████████████ 100.000% Copied 1000 records from 'mnist_training' (260.7 KB, 47.3 KB/s)   
```




### Version

1.13.0

### OS

Linux

### Snippet to repeat the bug

```
```

### Logs

```Shell
Dec 13 21:38:39 play run_reductstore.sh[104470]: 2024-12-13 21:38:39.744 (18752) [INFO] -- main.rs:23 ReductStore 1.13.0 [9c85c4f at 2024-12-04T16:36:46Z]
Dec 13 21:38:39 play run_reductstore.sh[104470]: 2024-12-13 21:38:39.745 (18752) [INFO] -- main.rs:38 Configuration:
Dec 13 21:38:39 play run_reductstore.sh[104470]:          RS_LOG_LEVEL = INFO (default)
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_HOST = 0.0.0.0 (default)
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_PORT = 443
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_API_BASE_PATH = / (default)
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_DATA_PATH = /mnt/data
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_API_TOKEN = ***********
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_CERT_PATH = *****************************************
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_CERT_KEY_PATH = *****************************************
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_LICENSE_PATH = /usr/local/lib/reductstore/license.lic
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_CORS_ALLOW_ORIGIN = *
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_BUCKET_1_NAME = demo
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_BUCKET_2_NAME = datasets
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_BUCKET_1_QUOTA_TYPE = FIFO
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_BUCKET_1_QUOTA_SIZE = 10.0 TB
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_TOKEN_1_NAME = dataset-read
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_TOKEN_2_NAME = demo
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_TOKEN_1_VALUE = *****************************************************************************
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_TOKEN_2_VALUE = ***********
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_TOKEN_1_READ = datasets
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_TOKEN_2_READ = demo,datasets
Dec 13 21:38:39 play run_reductstore.sh[104470]:         RS_TOKEN_2_WRITE = demo
Dec 13 21:38:40 play run_reductstore.sh[104470]: 2024-12-13 21:38:40.145 (31104) [WARN] -- storage/entry/entry_loader.rs:190 Recovering uncommitted changes from WALs for entry: "/mnt/data/demo/cats"
Dec 13 21:38:40 play run_reductstore.sh[104470]: 2024-12-13 21:38:40.245 (25408) [WARN] -- storage/entry/entry_loader.rs:190 Recovering uncommitted changes from WALs for entry: "/mnt/data/demo/imdb"
Dec 13 21:38:43 play run_reductstore.sh[104470]: 2024-12-13 21:38:43.766 (18752) [INFO] -- storage/storage.rs:67 Load 3 buckets
Dec 13 21:38:43 play run_reductstore.sh[104470]: 2024-12-13 21:38:43.776 (18752) [INFO] -- cfg.rs:146 Provisioned bucket 'datasets' with: BucketSettings { quota_type: Some(NONE), quota_size: Some(0), max_block_size: Some(64000000), max_block_records: Some(256) }
Dec 13 21:38:43 play run_reductstore.sh[104470]: 2024-12-13 21:38:43.786 (18752) [INFO] -- cfg.rs:146 Provisioned bucket 'demo' with: BucketSettings { quota_type: Some(FIFO), quota_size: Some(10000000000000), max_block_size: Some(64000000), max_block_records: Some(256) }
Dec 13 21:38:43 play run_reductstore.sh[104470]: 2024-12-13 21:38:43.793 (18752) [INFO] -- cfg.rs:112 Provisioned token 'dataset-read' with Some(Permissions { full_access: false, read: ["datasets"], write: [""] })
Dec 13 21:38:43 play run_reductstore.sh[104470]: 2024-12-13 21:38:43.793 (18752) [INFO] -- cfg.rs:112 Provisioned token 'demo' with Some(Permissions { full_access: false, read: ["demo", "datasets"], write: ["demo"] })
Dec 13 21:38:43 play run_reductstore.sh[104470]: 2024-12-13 21:38:43.793 (18752) [INFO] -- cfg.rs:372 Load Web Console
Dec 13 21:38:43 play run_reductstore.sh[104470]: 2024-12-13 21:38:43.826 (18752) [INFO] -- main.rs:46 License Information:
Dec 13 21:38:43 play run_reductstore.sh[104470]:         Licensee: Azure MarketPlace
Dec 13 21:38:43 play run_reductstore.sh[104470]:         Invoice: ---
Dec 13 21:38:43 play run_reductstore.sh[104470]:         Expiry Date: 2045-01-01 00:00:00 UTC
Dec 13 21:38:43 play run_reductstore.sh[104470]:         Plan: STANDARD
Dec 13 21:38:43 play run_reductstore.sh[104470]:         Number of Devices: 1
Dec 13 21:38:43 play run_reductstore.sh[104470]:         Disk Quota: 1 TB,
Dec 13 21:38:43 play run_reductstore.sh[104470]:         Fingerprint: 32063a5fd10e4741fa262f8bbc7bb21ce9a01eb57ae34005ec6f50838e718800
Dec 13 21:38:43 play run_reductstore.sh[104470]: 2024-12-13 21:38:43.826 (18752) [INFO] -- main.rs:60 Run HTTP reductstore on https://0.0.0.0:443/
Dec 13 21:38:58 play run_reductstore.sh[104470]: 2024-12-13 21:38:58.874 (18752) [ERROR] -- api/middleware.rs:56 POST /api/v1/b/demo/cats?ts=234 [500 Internal Server Error] Input/output error (os error 5)
Dec 13 21:39:01 play run_reductstore.sh[104470]: 2024-12-13 21:39:01.259 (18752) [ERROR] -- api/middleware.rs:56 POST /api/v1/b/demo/imdb?ts=332 [500 Internal Server Error] Input/output error (os error 5)
```


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Fixed
+
+- RS-555: Close WAL file before removal it from disk, [PR-706](https://github.com/reductstore/reductstore/pull/706)
+
 ## [1.13.2] - 2025-01-15
 
 ### Fixed

--- a/reductstore/src/core/file_cache.rs
+++ b/reductstore/src/core/file_cache.rs
@@ -1,15 +1,14 @@
 // Copyright 2023-2024 ReductSoftware UG
 // Licensed under the Business Source License 1.1
 
+use crate::core::cache::Cache;
+use reduct_base::error::ReductError;
+use reduct_base::internal_server_error;
 use std::fs::{remove_dir_all, remove_file, rename, File};
 use std::io::{Seek, SeekFrom};
 use std::path::PathBuf;
 use std::sync::{Arc, LazyLock, RwLock, Weak};
 use std::time::Duration;
-
-use crate::core::cache::Cache;
-use reduct_base::error::ReductError;
-use reduct_base::internal_server_error;
 
 const FILE_CACHE_MAX_SIZE: usize = 1024;
 const FILE_CACHE_TIME_TO_LIVE: Duration = Duration::from_secs(60);
@@ -177,12 +176,13 @@ impl FileCache {
     /// This function will return an error if the file does not exist or if there is an issue
     /// removing the file from the file system.
     pub fn remove(&self, path: &PathBuf) -> Result<(), ReductError> {
+        let mut cache = self.cache.write()?;
+        cache.remove(path);
+
         if path.try_exists()? {
             remove_file(path)?;
         }
-        let mut cache = self.cache.write()?;
 
-        cache.remove(path);
         Ok(())
     }
 

--- a/reductstore/src/storage/storage.rs
+++ b/reductstore/src/storage/storage.rs
@@ -49,7 +49,7 @@ impl Storage {
     /// If the data_path doesn't exist and can't be created, or if a bucket can't be restored.
     pub fn load(data_path: PathBuf, license: Option<License>) -> Storage {
         if !data_path.try_exists().unwrap_or(false) {
-            info!("Folder '{:?}' doesn't exist. Create it.", data_path);
+            info!("Folder {:?} doesn't exist. Create it.", data_path);
             std::fs::create_dir_all(&data_path).unwrap();
         }
 

--- a/reductstore/src/storage/storage.rs
+++ b/reductstore/src/storage/storage.rs
@@ -297,6 +297,14 @@ mod tests {
     use tempfile::tempdir;
 
     #[rstest]
+    fn test_create_folder(path: PathBuf) {
+        let path = path.join("test");
+        assert!(!path.exists());
+        let _ = Storage::load(path.clone(), None);
+        assert!(path.exists(), "Engine creates a folder if it doesn't exist");
+    }
+
+    #[rstest]
     fn test_info(storage: Storage) {
         sleep(Duration::from_secs(1)); // uptime is 1 second
 

